### ISSUE TITLE
Improve C++ Exception performance by adding new exception mode -- handle exception in native C++ only.

### ISSUE THREAD

**Issue by LanderlYoung** [emscripten-core/emscripten#12475]
Emscripten currently has two exception mode:
1. **disabled**: disable C++ exception completely.
2. **enabled**: enable exception and C++ exception can be throw to js

Here is a blog dive into the implementation [Exception handling in emscripten: how it works and why it’s disabled by default](https://brionv.com/log/2019/10/24/exception-handling-in-emscripten-how-it-works-and-why-its-disabled-by-default/)

The 2nd-way requires a lot of work to be done, such as frequent calls between WASM and JS, and that caused the performance issue.

So, my consideration is to **let C++ throw and catch exceptions by itself, never throw an exception from C++ to JS.**

To put it another way, **keep the exceptions awared only to C++**, thus we can use clang (libc++ and libc++abi) implementation to handle exceptions with high performance.

In order to implement such solution, the compiler config should be changed (add mode 3 for `DISABLE_EXCEPTION_CATCHING`, and the libc++abi for emscripten should be changed accordingly.

If anyone can confirm this solution, I'm pleased to open a PR.

**Comment by sbc100** [comment]
The problem is that for now there is no way to actually throw or catch from WebAssembly, which is why we need to use JS to add the try/cratch blocks.  

There is an exception handling proposal in WebAssembly https://github.com/WebAssembly/exception-handling/ and there is an experiment flag to enable this: `-fwasm-exceptions`.  Bare in mind that no major browser yet supports this standard so its not yet very useful.

Without the aforementioned proposal there is simply no way to throw or catch an exception from WebAssebmly so I'm afraid that what you are suggesting is not possible.

**Comment by LanderlYoung** [comment]
Thanks for the info.

It seems like the `-fwasm-exceptions` feature is not what I was expecting.

I just want wasm to support C++ code to work with exceptions, not throwing them to js (the embedder).

For instance:

```C++
// exported to the embedder
extern "C" void func() {
    // always handle C++ exceptions internally
    // so the embadder don't know any exceptions
    try {
        actions();
    } catch (const SomeCppException & e) { 
      handle(e);
   }
}
```

The `actions` function throws C++ exceptions inside, and I just want the C++ function to work correctly, without the embedder (JS code) to collaborate with C++/WASM.

To do this, I'll try just to use libc++abi's `__cxa_throw` implementation. (Emscripten currently have It's own `__cxa_throw` implemented in JS).

Anyway, I'll try to implement a POF first.

Thanks again.

**Comment by tlively** [comment]
You'll find that at some point libc++abi's `__cxa_throw` bottoms out with some sort of native assembly, which will be impossible to port to WebAssembly. There's no way around this until the Wasm EH proposal is standardized.

**Comment by kripken** [comment]
Yes, the crucial thing here is that to implement C++ exceptions you need to be able to unwind the stack. The wasm exceptions proposal will add that capability to wasm. Until then, we have to do something else.

Currently we do that unwinding using JS - see the `invoke` functions. They do a try-catch in JS for us, and we can catch a JS exception that represents a C++ exception.

There is another possible way - in theory wasm could emulate unwinding the same way Asyncify does, that is, "unwinding" sets a global "I am unwinding now", and all wasm code checks a global after each call, and if we are unwinding, it returns out, and so forth (credit: @lukewagner had this idea many years ago, long before Asyncify). Asyncify shows this is definitely possible, but I'd expect significant overhead to code size - not as bad as Asyncify, but much worse than the current JS support. So I am not very optimistic about that, but I could be wrong of course...

**Comment by LanderlYoung** [comment]
Thanks, guys. Learned a lot.

It seems it is indeed impossible to let C++ unwind stack for now.

One possible solution as @kripken mentioned is to add exception-checks all over the place, despite the code size concern, it also requires code generation to be changed.

But with the `-sDISABLE_EXCEPTION_CATCHING=0` option, the performance is really bad.

Hope the proposal be implemented in recent days.
